src/Refinery/Tactic.hs

```diff
@@ -125,18 +183,51 @@ ensure p f t = check >> t
           s <- get
           modify f
           case p s of
-            Just err -> throwError err
+            Just err -> unsolvable err
```

Comment:
Is there a difference between `unsolvable` and `unsolvableHole` and `failure`?

Reply:
Yeah, it's a matter of context.